patches/flashtools.patch

```diff
@@ -0,0 +1,302 @@
+ TARGETS += poke
+ TARGETS += peek
++TARGETS += cbfs
+ 
+ CFLAGS += \
+ 	-std=c99 \
+@@ -17,6 +18,7 @@ all: $(TARGETS)
+ flashtool: flashtool.o spiflash.o util.o
+ peek: peek.o util.o
+ poke: poke.o util.o
++cbfs: cbfs.o
+ 
+ $(TARGETS):
+ 	$(CC) $(LDFLAGS) -o $@ $^
+diff --git ./cbfs.c ./cbfs.c
+new file mode 100644
+index 0000000..7e11c88
+--- /dev/null
++++ ./cbfs.c
+@@ -0,0 +1,267 @@
++#include <stdio.h>
++#include <fcntl.h>
++#include <stdlib.h>
++#include <stdint.h>
++#include <string.h>
++#include <inttypes.h>
++#include <errno.h>
++#include <unistd.h>
++#include <getopt.h>
++#include <arpa/inet.h>
++
++#define CBFS_HEADER_MAGIC  0x4F524243
++#define CBFS_HEADER_VERSION1 0x31313131
++#define CBFS_HEADER_VERSION2 0x31313132
++#define CBFS_HEADER_VERSION  CBFS_HEADER_VERSION2
++
++int verbose = 0;
++
++static const struct option long_options[] = {
++	{ "verbose",		0, NULL, 'v' },
++	{ "read",		1, NULL, 'r' },
++	{ "list",		0, NULL, 'l' },
++	{ "help",		0, NULL, 'h' },
++	{ NULL,			0, NULL, 0 },
++};
++
++
++static const char usage[] =
++"Usage: sudo cbfs [options]\n"
++"\n"
++"-h | -? | --help       This help\n"
++"-v | --verbose         Increase verbosity\n"
++"-r | --read file       Read the CBFS filed and dump to stdout\n"
++"-l | --list            List the raw files in the CBFS\n"
++"\n";
++
++struct cbfs_header {
++	uint32_t magic;
++	uint32_t version;
++	uint32_t romsize;
++	uint32_t bootblocksize;
++	uint32_t align; /* hard coded to 64 byte */
++	uint32_t offset;
++	uint32_t architecture;  /* Version 2 */
++	uint32_t pad[1];
++};
++
++#define CBFS_FILE_MAGIC "LARCHIVE"
++
++struct cbfs_file {
++	uint8_t magic[8];
++	/* length of file data */
++	uint32_t len;
++	uint32_t type;
++	/* offset to struct cbfs_file_attribute or 0 */
++	uint32_t attributes_offset;
++	/* length of header incl. variable data */
++	uint32_t offset;
++	char filename[];
++};
++
++
++int main(int argc, char** argv) {
++	const char * const prog_name = argv[0];
++	if (argc <= 1)
++	{
++		fprintf(stderr, "%s", usage);
++		return EXIT_FAILURE;
++	}
++
++	int opt;
++	int do_read = 0;
++	int do_list = 0;
++	const char * filename = NULL;
++	while ((opt = getopt_long(argc, argv, "h?vlr:", long_options, NULL)) != -1)
++	{
++		switch(opt)
++		{
++		case 'v':
++			verbose++;
++			break;
++		case 'l':
++			do_list = 1;
++			break;
++		case 'r':
++			do_read = 1;
++			filename = optarg;
++			break;
++		case '?': case 'h':
++			printf("%s", usage);
++			return EXIT_SUCCESS;
++		default:
++			fprintf(stderr, "%s", usage);
++			return EXIT_FAILURE;
++		}
++	}
++
++	argc -= optind;
++	argv += optind;
++	if (argc != 0)
++	{
++		fprintf(stderr, "%s: Excess arguments?\n", prog_name);
++		return EXIT_FAILURE;
++	}
++
++	int fd;
++	fd = open("/dev/mem", O_RDONLY);
++	if (fd < 0) {
++		fprintf(stderr, "Failed to open /dev/mem : %s\n", strerror(errno));
++		return EXIT_FAILURE;
++	}
++
++	uint64_t end = 0x100000000;
++	if (verbose) {
++		printf("Seeking to %lx\n", end-4);
```

Comment:
Since you print the file to stdout, all the verbose printfs should probably go to sdterr.